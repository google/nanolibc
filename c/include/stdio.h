//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_

#include <stdarg.h>
#include <sys/types.h>
#include <sys/_cdefs.h>

/* Some assume stdio.h defines NULL, etc. And SUSv2 is actually ok with
 * that: "Inclusion of the <stdio.h> header may also make visible all
 * symbols from <stddef.h>." */
#include <stddef.h>

__BEGIN_DECLS

#define EOF (-1)

typedef struct _FILE_DESCR FILE;

extern FILE *stdout;
extern FILE *stderr;

int vsnprintf(char *str, size_t size, const char *format, va_list ap);
int snprintf(char *str, size_t size, const char *format, ...)
    __attribute__((format(printf, 3, 4)));

int vsprintf(char *str, const char *format, va_list ap);
int sprintf(char *str, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vfprintf(FILE *stream, const char *format, va_list ap);
int fprintf(FILE *stream, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vprintf(const char *format, va_list ap);
int printf(const char *format, ...) __attribute__((format(printf, 1, 2)));

int fputc(int c, FILE *stream);
int fputs(const char *s, FILE *stream);
int putc(int c, FILE *stream);
int putchar(int c);
int puts(const char *s);

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);

int fflush(FILE *stream);

int perror(const char *s);

/*
 * Published but functions NOT implemented!
 */
extern FILE *stdin;

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

int scanf(const char *format, ...);
int sscanf(const char *str, const char *format, ...);
FILE *fopen(const char *path, const char *mode);
int fclose(FILE *fp);
char *fgets(char *s, int size, FILE *stream);
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
int fseek(FILE *stream, long offset, int whence);
long ftell(FILE *stream);
void rewind(FILE *stream);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
