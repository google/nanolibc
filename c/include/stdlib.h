//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STDLIB_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STDLIB_H_

#include <stddef.h>
#include <sys/types.h>
#include <malloc.h> /* malloc/free */

#include <sys/_cdefs.h>

#define EXIT_FAILURE 1
#define EXIT_SUCCESS 0

__BEGIN_DECLS

long int strtol(const char *nptr, char **endptr, int base);
long long int strtoll(const char *nptr, char **endptr, int base);
unsigned long int strtoul(const char *nptr, char **endptr, int base);
unsigned long long int strtoull(const char *nptr, char **endptr, int base);
long atol(const char *nptr);

float strtof(const char *nptr, char **endptr);
double strtod(const char *nptr, char **endptr);

void exit(int status) __attribute__((noreturn));
void abort(void) __attribute__((noreturn));

/*
 * Published but functions NOT implemented!
 */
char *getenv(const char *name);
void qsort(void *base, size_t nmemb, size_t size,
           int (*compar)(const void *, const void *));

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STDLIB_H_
