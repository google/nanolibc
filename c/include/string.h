//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_

#include <sys/_cdefs.h>
#include <sys/types.h>

__BEGIN_DECLS

int memcmp(const void *s1, const void *s2, size_t n);
void *memcpy(void *dstpp, const void *srcpp, size_t n);
void *memmove(void *dstpp, const void *srcpp, size_t n);
void *memset(void *dstpp, int c, size_t n);
void *memchr(const void *spp, int c, size_t n);
const void *rawmemchr(const void *spp, int c);

int strcmp(const char *p1, const char *p2);
char *strcpy(char *dest, const char *src);
size_t strlen(const char *s);
int strncmp(const char *s1, const char *s2, size_t n);
char *strncpy(char *dest, const char *src, size_t n);
char *strchr(const char *s, int i);
char *strrchr(const char *s, int i);
const char *strstr(const char *haystack, const char *needle);
size_t strspn(const char *s, const char *accept);
const char *strpbrk(const char *s, const char *accept);

// These two are unsupported.
char *strtok_r(char *str, const char *delim, char **saveptr);
char *strtok(char *str, const char *delim);

char *strdup(const char *s);
char *strndup(const char *s, size_t n);

char *strerror(int errnum);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_
