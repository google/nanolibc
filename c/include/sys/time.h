//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_SYS_TIME_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_SYS_TIME_H_

#include <sys/_cdefs.h>
#include <sys/types.h>

__BEGIN_DECLS

struct timeval {
  time_t      tv_sec;     /* seconds */
  suseconds_t tv_usec;    /* microseconds */
};

int gettimeofday(struct timeval *tv, void *unsupported_tz);

/*
 * _BSD_SOURCE extensions
 *
 * All functions below expect timeval params to be normalized.
 */
#define _GGL_USECS_PER_SEC (1000000L)

inline static void timeradd(struct timeval *a, struct timeval *b,
                            struct timeval *res) {
  res->tv_sec = a->tv_sec + b->tv_sec;
  res->tv_usec = a->tv_usec + b->tv_usec;
  if (res->tv_usec >= _GGL_USECS_PER_SEC) {
    res->tv_sec++;
    res->tv_usec -= _GGL_USECS_PER_SEC;
  }
}

inline static void timersub(struct timeval *a, struct timeval *b,
                            struct timeval *res) {
  res->tv_sec = a->tv_sec - b->tv_sec;
  res->tv_usec = a->tv_usec - b->tv_usec;
  if (res->tv_usec < 0) {
    res->tv_sec--;
    res->tv_usec += _GGL_USECS_PER_SEC;
  }
}

#define timercmp(a, b, CMP)           \
  (((a)->tv_sec == (b)->tv_sec) ?     \
   ((a)->tv_usec CMP (b)->tv_usec) :  \
   ((a)->tv_sec CMP (b)->tv_sec))

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_SYS_TIME_H_
