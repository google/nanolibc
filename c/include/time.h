//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_TIME_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_TIME_H_

#include <sys/_cdefs.h>
#include <sys/types.h>

__BEGIN_DECLS

struct tm {
  int tm_sec;
  int tm_min;
  int tm_hour;
  int tm_mday;
  int tm_mon;
  int tm_year;
  int tm_wday;
  int tm_yday;
  int tm_isdst;
  long int tm_gmtoff;
  const char *tm_zone;
};

struct timespec {
  time_t   tv_sec;        /* seconds */
  long     tv_nsec;       /* nanoseconds */
};

#define CLOCK_REALTIME  0
#define TIMER_ABSTIME   1
#define CLOCK_MONOTONIC 1

time_t time(time_t *t);
int clock_gettime(clockid_t clk_id, struct timespec *tp);

/*
 * Published but functions NOT implemented!
 */
struct tm *localtime(const time_t *timep);

// Not implemented.
unsigned int clock(void);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_TIME_H_
