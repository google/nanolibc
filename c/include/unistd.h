//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_UNISTD_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_UNISTD_H_

#include <sys/_cdefs.h>
#include <sys/types.h>

__BEGIN_DECLS

unsigned int usleep(unsigned int useconds);
unsigned int sleep(unsigned int seconds);
ssize_t write(int fd, const void *buf, size_t count);

#define STDIN_FILENO 0
#define STDOUT_FILENO 1
#define STDERR_FILENO 2

/*
 * Published but functions NOT implemented!
 */
int isatty(int fd);
int getopt(int argc, char * const argv[], const char *optstring);
ssize_t read(int fd, void *buf, size_t count);

/*
 * Global variables.
 */
/* for getopt: */
extern char *optarg;
extern int optind, opterr, optopt;

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_UNISTD_H_
