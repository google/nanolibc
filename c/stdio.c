//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of stdio.h functions.
 */

#include <stddef.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>

#include "third_party/nanolibc/c/libc_internals.h"
#include "third_party/tinyprintf/tinyprintf.h"

FILE *stdin = &__nnlc_internal_data.libc_stdin;
FILE *stdout = &__nnlc_internal_data.libc_stdout;
FILE *stderr = &__nnlc_internal_data.libc_stderr;

/* returns TRUE if param 'stream' is a valid FILE* pointer */
inline static int _is_valid_FILE(FILE *stream) {
  return ((stream != NULL) && (stream->magic == _NNLC_STDIO_MAGIC));
}

int fputc(int c, FILE *stream) {
  const unsigned char uc = c; /* according to man */
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  if (1 != stream->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int fputs(const char *s, FILE *stream) {
  size_t len;
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  len = strlen(s);
  if (stream->write(s, len) != (ssize_t)len) return EOF;
  return 1;
}

int putc(int c, FILE *stream) { return fputc(c, stream); }

int putchar(int c) {
  const unsigned char uc = c; /* according to man */
  if (1 != stdout->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int puts(const char *s) {
  const size_t len = strlen(s);
  if (stdout->write(s, len) != (ssize_t)len) return EOF;
  if (1 != stdout->write("\n", 1)) return EOF;
  return 1;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  size_t i;
  if (!_is_valid_FILE(stream) || !stream->write) return 0;
  for (i = 0; i < nmemb; ++i) {
    const void *p = ptr + (i * size);
    if (stream->write(p, size) != (ssize_t)size) break;
  }
  return i;
}

int fflush(FILE *stream) {
  if (!_is_valid_FILE(stream)) return -1;

  return 0;
}

int perror(const char *s) {
  fprintf(stderr, "Error: %s\n", s);
  return 0;
}

/*
 * vprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to call the underlying runtime 'write'
 * functions. We use a small buffer on the stack to avoid calling the
 * underlying 'write' too often.
 */
struct _vprintf_putcf_data {
  FILE *stream;

  char *buff; /* allocated on the stack by vprintf */
  size_t buff_capacity;
  size_t nbuffered; /* slightly redundant with buff_capacity, we don't care */

  size_t total_written;
};

/* Callback function for tfp_format to add the given char c to the buffer */
static void _vprintf_putcf(void *p, char c) {
  struct _vprintf_putcf_data *data = (struct _vprintf_putcf_data *)p;

  /* sanity check */
  if (data->nbuffered >= data->buff_capacity)
    abort(); /* no assert(), to prevent infinite recursion */

  data->buff[data->nbuffered++] = c;

  /* need to flush buffer? */
  if (data->nbuffered >= data->buff_capacity) {
    int n = data->stream->write(data->buff, data->nbuffered);
    if (n > 0) data->total_written += n;
    data->nbuffered = 0;
  }
}

int vfprintf(FILE *stream, const char *format, va_list ap) {
  char buff[64]; /* any size >= 1 is Ok */
  struct _vprintf_putcf_data data;

  if (!_is_valid_FILE(stream) || !stream->write) return -1;

  data.stream = stream;
  data.buff_capacity = sizeof(buff);
  data.buff = buff;
  data.nbuffered = 0;
  data.total_written = 0;
  tfp_format(&data, _vprintf_putcf, format, ap);

  /* chars remaining in buffer? */
  if (data.nbuffered > 0) {
    int n = stream->write(data.buff, data.nbuffered);
    if (n > 0) data.total_written += n;
  }

  return data.total_written;
}

int fprintf(FILE *stream, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vfprintf(stream, format, ap);
  va_end(ap);
  return retval;
}

int vprintf(const char *format, va_list ap) {
  return vfprintf(stdout, format, ap);
}

int printf(const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vprintf(format, ap);
  va_end(ap);

  return retval;
}

/*
 * vsnprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsnprintf_putcf_data {
  size_t dest_capacity;
  char *dest;
  size_t dest_used;
};

static void _vsnprintf_putcf(void *p, char c) {
  struct _vsnprintf_putcf_data *data = (struct _vsnprintf_putcf_data *)p;
  if (data->dest_used < data->dest_capacity) data->dest[data->dest_used] = c;

  data->dest_used++;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {
  struct _vsnprintf_putcf_data data;

  data.dest = str;
  data.dest_capacity = size;
  data.dest_used = 0;
  tfp_format(&data, _vsnprintf_putcf, format, ap);

  if (data.dest_used < size)
    data.dest[data.dest_used] = '\0';
  else if (size > 0)
    data.dest[size - 1] = '\0';

  return data.dest_used;
}

int snprintf(char *str, size_t size, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsnprintf(str, size, format, ap);
  va_end(ap);
  return retval;
}

/*
 * vsprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsprintf_putcf_data {
  char *dest;
  size_t dest_used;
};

static void _vsprintf_putcf(void *p, char c) {
  struct _vsprintf_putcf_data *data = (struct _vsprintf_putcf_data *)p;
  data->dest[data->dest_used++] = c;
}

int vsprintf(char *str, const char *format, va_list ap) {
  struct _vsprintf_putcf_data data;
  data.dest = str;
  data.dest_used = 0;
  tfp_format(&data, _vsprintf_putcf, format, ap);
  data.dest[data.dest_used] = '\0';
  return data.dest_used;
}

int sprintf(char *str, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsprintf(str, format, ap);
  va_end(ap);
  return retval;
}
