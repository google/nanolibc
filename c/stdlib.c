//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of stdlib.h/unistd.h functions.
 */

#include <limits.h>
#include <ctype.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "third_party/nanolibc/c/libc_internals.h"

void abort() { exit(-1); }

void exit(int status) {
  __nnlc_internal_data.sysdeps->exit(status);

  /* runtime's exit() should not return */
  fprintf(stderr, "exit() failed!!! Infinite loop...\n");
  while (1) continue;
}

unsigned int sleep(unsigned int seconds) {
  return __nnlc_internal_data.sysdeps->usleep(seconds * 1000000ULL);
}

unsigned int usleep(unsigned int useconds) {
  return __nnlc_internal_data.sysdeps->usleep(useconds);
}

/*
 * string -> int conversions
 *
 * for all strtol variants, we use the same template source file
 */

#define _STRTOX_FNAME strtol
#define _STRTOX_SIGNED_TNAME long int
#define _STRTOX_VARIANT_UNSIGNED 0
#define _STRTOX_INTMIN LONG_MIN
#define _STRTOX_INTMAX LONG_MAX
#include "strtox.c"
#undef _STRTOX_FNAME
#undef _STRTOX_SIGNED_TNAME
#undef _STRTOX_VARIANT_UNSIGNED
#undef _STRTOX_INTMIN
#undef _STRTOX_INTMAX

#define _STRTOX_FNAME strtoll
#define _STRTOX_SIGNED_TNAME long long int
#define _STRTOX_VARIANT_UNSIGNED 0
#define _STRTOX_INTMIN LLONG_MIN
#define _STRTOX_INTMAX LLONG_MAX
#include "strtox.c"
#undef _STRTOX_FNAME
#undef _STRTOX_SIGNED_TNAME
#undef _STRTOX_VARIANT_UNSIGNED
#undef _STRTOX_INTMIN
#undef _STRTOX_INTMAX

#define _STRTOX_FNAME strtoul
#define _STRTOX_SIGNED_TNAME long int
#define _STRTOX_VARIANT_UNSIGNED 1
#define _STRTOX_INTMIN ULONG_MIN
#define _STRTOX_INTMAX ULONG_MAX
#include "strtox.c"
#undef _STRTOX_FNAME
#undef _STRTOX_SIGNED_TNAME
#undef _STRTOX_VARIANT_UNSIGNED
#undef _STRTOX_INTMIN
#undef _STRTOX_INTMAX

#define _STRTOX_FNAME strtoull
#define _STRTOX_SIGNED_TNAME long long int
#define _STRTOX_VARIANT_UNSIGNED 1
#define _STRTOX_INTMIN ULLONG_MIN
#define _STRTOX_INTMAX ULLONG_MAX
#include "strtox.c"
#undef _STRTOX_FNAME
#undef _STRTOX_SIGNED_TNAME
#undef _STRTOX_VARIANT_UNSIGNED
#undef _STRTOX_INTMIN
#undef _STRTOX_INTMAX

long int atol(const char *nptr) { return strtol(nptr, NULL, 10); }
