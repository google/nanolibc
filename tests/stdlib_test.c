//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* test stdlib.h/unistd.h */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <time.h>
#include <unistd.h>

#include "third_party/nanolibc/tests/test_utils/nnlc_test.h"

#if defined(NNLC_MALLOC)
  #define ALIGNMENT NNLC_MALLOC_ALIGNMENT
#elif defined(__clang__) && defined(__x86_64)
  #define ALIGNMENT ((size_t)16)
#else
  #define ALIGNMENT ((size_t)1)
#endif

static void test_malloc_free() {
  char *p;
  int i;

  printf("Starting malloc/free tests...\n");

  free(NULL); /* should not crash */

  p = malloc(0);
  /* value of p is undefined */
  if (p != NULL) free(p);

  p = malloc(1);
  ASSERT(p != NULL);
  ASSERT(((intptr_t)p & (ALIGNMENT - 1)) == 0);
  *p = 'X';
  ASSERT(*p == 'X');
  free(p);

  p = realloc(NULL, 0);
  /* value of p is undefined */
  if (p != NULL) free(p);

  p = realloc(NULL, 1);
  ASSERT(p != NULL);
  *p = 'X';
  ASSERT(*p == 'X');
  p = realloc(p, 17);
  ASSERT(p != NULL);
  p[0] = 'X';
  p[16] = 'Z';
  ASSERT(p[0] == 'X');
  ASSERT(p[16] == 'Z');

  /* Semantic of realloc(3) for size 0 is implementation-specific. */
  {
    void *q = realloc(p, 0);
    if (q != NULL) {
      /* What to do with p (freeing or not freeing it) is not clear:
       * specs seem to differ slightly.
       * In one interpretation, we could:
       *   ASSERT(q == p);
       *   free(p);
       * But I prefer not to assume that, considering the specs are somewhat
       * vague.
       * And in C23, this becomes undefined behavior according to
       * https://en.cppreference.com/w/c/memory/realloc
       */
    }
  }

  for (i = 4; i < 5 * 1024 * 1024; i += (64 + i / 100)) {
    const int realloc_sz = 5 * 1024 * 1024 - i;
    p = malloc(i);
    ASSERT(p != NULL);
    ASSERT(((intptr_t)p & (ALIGNMENT - 1)) == 0);
    memset(p, 0x5a, i);
    *p = 'X';
    p[i - 1] = 'Z';
    ASSERT(*p == 'X');
    ASSERT(p[1] == 0x5a);
    ASSERT(p[i - 2] == 0x5a);
    ASSERT(p[i - 1] == 'Z');

    p = realloc(p, realloc_sz);
    if (i <= realloc_sz) {
      *p = 'X';
      p[i - 1] = 'Z';
      ASSERT(*p == 'X');
      ASSERT(p[1] == 0x5a);
      ASSERT(p[i - 2] == 0x5a);
      ASSERT(p[i - 1] == 'Z');
    } else {
      ASSERT(*p == 'X');
      ASSERT(p[1] == 0x5a);
      ASSERT(p[realloc_sz - 1] == 0x5a);
    }

    free(p);
  }

  printf("Done with malloc/free tests.\n");
}

static void test_calloc() {
  char *buf = calloc(3, sizeof(int));
  ASSERT(!memcmp(buf, "\x00\x00\x00", 3));
  buf[0] = 1;
  buf[1] = 2;
  buf[2] = 3;
  ASSERT(buf[0] == 1 && buf[1] == 2 && buf[2] == 3);
#ifdef NNLC_MALLOC
  // It turns out the old calloc in crosstool fails to catch the overflow.  This
  // has been fixed in newer versions of gcc's libc.
  size_t large = ((size_t)-1) >> 5;
  size_t many = ((size_t)-1) >> 4;
  void* buf2 = calloc(many, large);
  ASSERT(buf2 == NULL);
#endif
  free(buf);
}

static int timespec_cmp(const struct timespec *a,
                        const struct timespec *b) {
  if (a->tv_sec < b->tv_sec)
    return -1;
  else if (a->tv_sec > b->tv_sec)
    return 1;
  else if (a->tv_nsec < b->tv_nsec)
    return -1;
  else if (a->tv_nsec > b->tv_nsec)
    return 1;
  return 0;
}

static void test_sleep() {
  int i;
  time_t t1, t2;
  struct timespec tp1, tp2;
  struct timeval tv1, tv2, tvdiff, tv2check;

  printf("sleep(0)...\n");
  sleep(0);
  printf("done. sleep(1)...\n");
  sleep(1);
  printf("done. sleep(3)...\n");
  sleep(3);
  printf("done. 10 * usleep(200*1000)...\n");
  ASSERT(0 == clock_gettime(CLOCK_MONOTONIC, &tp1));
  ASSERT(0 == gettimeofday(&tv1, NULL));
  t1 = time(NULL);
  for (i = 0; i < 10; ++i) usleep(200 * 1000);  /* ~ 2s */
  ASSERT(0 == clock_gettime(CLOCK_MONOTONIC, &tp2));
  ASSERT(0 == gettimeofday(&tv2, NULL));
  time(&t2);
  /* the following will work only if we are guaranteed that tp2 > tp1
   * + 1s on EFI (clock resolution is 1s). */
  printf("start time was: tp=%ld.%09lds tv=%ld.%06lds\n",
         tp1.tv_sec, tp1.tv_nsec,
         tv1.tv_sec, tv1.tv_usec);
  timersub(&tv2, &tv1, &tvdiff);
  printf("end time: %ld.%09lds tv=%ld.%06lds (tvdiff=%ld.%06lds)\n",
         tp2.tv_sec, tp2.tv_nsec,
         tv2.tv_sec, tv2.tv_usec,
         tvdiff.tv_sec, tvdiff.tv_usec);
  ASSERT(timespec_cmp(&tp1, &tp2) < 0);
  ASSERT(timercmp(&tv1, &tv2, <));
  ASSERT(t2 >= t1 + 1);
  /* check that timeradd(t1, timesub(t2, t1)) == t2 */
  timeradd(&tv1, &tvdiff, &tv2check);
  ASSERT(timercmp(&tv2check, &tv2, ==));
  printf("done.\n");
}

int main() {
  test_malloc_free();
  test_calloc();
  test_sleep();

  printf("That's all, folks!\n");
  return 0;
}
