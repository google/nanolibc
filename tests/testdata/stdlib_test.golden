Starting malloc/free tests...
Done with malloc/free tests.
sleep(0)...
done. sleep(1)...
done. sleep(3)...
done. 10 * usleep(200*1000)...
start time was: REDACTED
end time: REDACTED
done.
That's all, folks!
# END TEST WITH RETVAL=0
